tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp

```diff
@@ -0,0 +1,705 @@
+            auto logical_idx_start = idx_pair[0];
+            auto physical_idx_start = idx_pair[1];
+
+            for (size_t col = 0; col < cols; col++) {
+                data[logical_idx_start + col] = physical_data[physical_idx_start + col];
+            }
+        }
+    }
+
+    TT_FATAL(
+        data.size() == shape.volume(),
+        "Data size {} should be same as volume indicated by shape {}",
+        data.size(),
+        shape);
+
+    return data;
+};
+
+}  // namespace
+
+class ShardWithAlignmentTests : public ::testing::TestWithParam<ShardWithAlignmentParams> {};
+
+TEST_P(ShardWithAlignmentTests, LogicalToPhysical) {
+    const auto& params = GetParam();
+
+    auto physical_shard_height = tt::round_up(params.inputs.shard_shape.height(), params.inputs.shard_alignment[0]);
+    auto physical_shard_width = tt::round_up(params.inputs.shard_shape.width(), params.inputs.shard_alignment[1]);
+    Size physical_shard_shape{physical_shard_height, physical_shard_width};
+    ASSERT_EQ(physical_shard_shape, params.expected.physical_shard_shape);
+
+    auto physical_size = get_physical_shape(params.inputs.shape, params.inputs.shard_shape, physical_shard_shape);
+    ASSERT_EQ(physical_size, params.expected.physical_size);
+
+    const auto& data = params.inputs.data;
+    const auto& expected_physical_data = params.expected.physical_data;
+
+    auto physical_data = convert_fp32_logical_data_to_physical_data(
+        data, params.inputs.shape, params.inputs.shard_shape, physical_shard_shape);
+
+    // auto shape_2D = flatten_to_2D(params.inputs.shape);
+    // pretty_print_data_as_shards(data, shape_2D, params.inputs.shard_shape);
+    // pretty_print_data_as_shards(physical_data, physical_size, physical_shard_shape);
+
+    ASSERT_EQ(physical_data.size(), expected_physical_data.size());
+    for (size_t i = 0; i < physical_data.size(); i++) {
+        EXPECT_EQ(physical_data[i], expected_physical_data[i]);
+    }
+}
+
+TEST_P(ShardWithAlignmentTests, PhysicalToLogical) {
+    const auto& params = GetParam();
+
+    auto physical_shard_height = tt::round_up(params.inputs.shard_shape.height(), params.inputs.shard_alignment[0]);
+    auto physical_shard_width = tt::round_up(params.inputs.shard_shape.width(), params.inputs.shard_alignment[1]);
+    Size physical_shard_shape{physical_shard_height, physical_shard_width};
+    ASSERT_EQ(physical_shard_shape, params.expected.physical_shard_shape);
+
+    auto physical_size = get_physical_shape(params.inputs.shape, params.inputs.shard_shape, physical_shard_shape);
+    ASSERT_EQ(physical_size, params.expected.physical_size);
+
+    // Use expected value as input physical data
+    const auto& physical_data = params.expected.physical_data;
+    const auto& expected_data = params.inputs.data;
+
+    auto data = convert_fp32_physical_data_to_logical_data(
+        physical_data, params.inputs.shape, params.inputs.shard_shape, physical_shard_shape);
+
```

Comment:
Remove?

Reply:
This shows you how to use `pretty_print_data_as_shards` which is really useful for debug.